Feature/go snippets

sacOO7 opened the pull request:
- Added go snippets

sacOO7 committed:
Created multiple code snippet files

sacOO7 committed:
Added code snippets for go

lmars commented on go/chat_pub.go:
```diff
@@ -0,0 +1,12 @@
+client, err := ably.NewRealtime(ably.WithKey("ABLY_API_KEY"))
+channel := client.Channels.Get('room:tesla-fans')
+pErr := channel.Presence.Enter(context.Background(), UserName + " entered the channel")
```
Let's just call this `err`

lmars commented on go/chat_pub.go:
```diff
@@ -0,0 +1,12 @@
+client, err := ably.NewRealtime(ably.WithKey("ABLY_API_KEY"))
```
Let's follow this with:

```go
if err != nil {
	return err
}
```

lmars commented on go/chat_pub.go:
```diff
@@ -0,0 +1,12 @@
+pErr := channel.Presence.Enter(context.Background(), UserName + " entered the channel")
+if pErr != nil {
+	err := fmt.Errorf("error with enter presence on the channel %w", pErr)
+	fmt.Println(err)
```
Let's just imagine this is wrapped in a function with a single `error` return value, and thus do:

```go
if err != nil {
	return fmt.Errorf("error entering presence on the channel %w", err)
}
```

sacOO7 replied:
Removed all error handling, since it's a unnecessary distrction for someone who reads the code and also its not present in any other code snippets

lmars commented on go/chat_pub.go:
```diff
@@ -0,0 +1,12 @@
+client, err := ably.NewRealtime(ably.WithKey("ABLY_API_KEY"))
+channel := client.Channels.Get('room:tesla-fans')
```
This string should be double quoted.

lmars commented on go/chat_pub.go:
```diff
@@ -0,0 +1,12 @@
+	err := fmt.Errorf("error with enter presence on the channel %w", pErr)
+	fmt.Println(err)
+}
+err = channel.Publish(context.Background(), 'msg', "{'user': 'elon', 'message': 'Hello fans'}")
```
Let's pass a `map[string]string` that will get serialised by the library rather than publishing a JSON string directly (since it will have the incorrect encoding), also also double quote `msg`:

```go
err = channel.Publish(context.Background(), "msg", map[string]string{"user": "elon", "message": "Hello fans"})
if err != nil {
	return fmt.Errorf("error publishing to the channel: %w", err)
}
```

lmars commented on go/chat_sub.go:
```diff
@@ -0,0 +1,25 @@
+		fmt.Printf("%v has left the chat\n", msg.ClientID)
+	}
+})
+defer unsubscribeAllPresence()
```
This `defer` should come after the check of `err`

lmars commented on go/chat_sub.go:
```diff
@@ -0,0 +1,25 @@
+	fmt.Printf("Received message from %v: '%v'\n", msg.ClientID, msg.Data)
+})
+
+defer unsubscribeAll()
```
This `defer` should come after the check of `err`

lmars reviewed (commented):
I'm definitely missing the context of what this code is for, but I've assumed it's some example code and commented accordingly

sacOO7 committed:
Updated code as per comments

jamienewcomb commented:
@sacOO7 did you translate the JS examples as requested? These look considerably longer than what we have in JS

sacOO7 commented:
@jamienewcomb we need to specify the error handling explicitly, it's bit different than what we have in other languages.

jamienewcomb commented:
@sacOO7 Okay I will let Lewis comment on this

Can we please remove the API key bits. We do not need these as part of these examples. Please refer to the JS ones

Edit: Lewis is going to comment soon

sacOO7 committed:
Removed distracting error handling, updated wrt js client

sacOO7 committed:
Removed unnecessary client returned from newRealtime object

sacOO7 commented:
@jamienewcomb I have removed client creation code snippet containing API keys. 
Current code is as per JS convention. (u can undertand stuff by looking at it, can't copy paste it directly)
Also, I had to remove ui context calls, since go natively doesn't support any UI framework.

sacOO7 committed:
Updated map at places where hybrid values being sent

lmars commented:
@sacOO7 apologies there's been some confusion around which Go snippets are required.

I've updated [the Jira ticket](https://ably.atlassian.net/browse/SDK-885) with an explicit list, which I'll also copy here:

The Go snippets needed are:

- Chat (publish)
- GPS (publish)
- IoT (publish and subscribe)
- Live Updates (publish)
- Multi-user Spaces (publish)

Please drop any snippets you've added which aren't in that list, and please bring the remaining ones inline with [the JavaScript snippets](https://github.com/ably/homepage-snippets/tree/main/javascript).

sacOO7 committed:
Removed unnecessary spec items as per ticket

sacOO7 commented:
> @sacOO7 apologies there's been some confusion around which Go snippets are required.
> 
> I've updated [the Jira ticket](https://ably.atlassian.net/browse/SDK-885) with an explicit list, which I'll also copy here:
> 
> The Go snippets needed are:
> 
> * Chat (publish)
> * GPS (publish)
> * IoT (publish and subscribe)
> * Live Updates (publish)
> * Multi-user Spaces (publish)
> 
> Please drop any snippets you've added which aren't in that list, and please bring the remaining ones inline with [the JavaScript snippets](https://github.com/ably/homepage-snippets/tree/main/javascript).

Updated

lmars commented on go/chat_pub.go:
```diff
@@ -0,0 +1,3 @@
+channel := client.Channels.Get("room:tesla-fans")
+channel.Presence.Enter(context.Background(), UserName + " entered the channel")
```
Please use `ctx` rather than `context.Background()` to save some horizontal space

lmars commented on go/chat_pub.go:
```diff
@@ -0,0 +1,3 @@
+channel := client.Channels.Get("room:tesla-fans")
+channel.Presence.Enter(context.Background(), UserName + " entered the channel")
+channel.Publish(context.Background(), "msg", map[string]string{"user": "elon", "message": "Hello fans"})
```
Please split this over two lines:

```go
msg := map[string]string{"user": "elon", "message": "Hello fans"}
channel.Publish(ctx, "msg", msg)
```

lmars commented on go/gps_pub.go:
```diff
@@ -0,0 +1,2 @@
+channel := client.Channels.Get("driver:elon")
+err = channel.Publish(context.Background(), "position", map[string]interface{}{ "lat": 65, "long": 98, "bearing": "nw" })
```
Please split this over two lines:

```go
msg := map[string]interface{}{"lat": lat, "long": long, "bearing": "nw"}
channel.Publish(ctx, "position", msg)
```

lmars commented on go/iot_sub.go:
```diff
@@ -0,0 +1,4 @@
+// Subscribe to messages sent on the channel with given eventName
```
This snippet needs to use an MQTT client not an Ably client (see the [equivalent JavaScript snippet](https://github.com/ably/homepage-snippets/blob/main/javascript/iot-sub.js)), and please drop the comment

lmars commented on go/live_updates_pub.go:
```diff
@@ -0,0 +1,2 @@
+channel := client.Channels.Get("dogecoin:usd")
+channel.Publish(context.Background(), "rate", map[string]int{ "price": 2, "time": 543564332 })
```
Please split this over two lines, and use variables in the message like [the JavaScript snippet](https://github.com/ably/homepage-snippets/blob/main/javascript/live-updates-pub.js)

lmars commented on go/multi_user_pub.go:
```diff
@@ -0,0 +1,2 @@
+channel := client.Channels.Get("doc:mars-launch-plan")
+channel.Publish(context.Background(), "update", map[string]string{ "pos": "charAt", "insert": "Send Grimes first" })
```
Please split this over two lines, and make `charAt` a variable reference rather than a string

lmars commented on go/iot_pub.go:
```diff
@@ -0,0 +1,2 @@
+channel := client.Channels.Get("device:rocket:1")
+channel.Publish(context.Background(), "action", map[string]string{ "task": "enable-thrusters" })
```
Please add a trailing newline to all source files

lmars committed:
Update Go snippets

Signed-off-by: Lewis Marshall <lewis.marshall@ably.com>

lmars commented:
I've pushed some updated based on my most recent review, I'm going to merge this, thanks @sacOO7
